src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -657,55 +658,67 @@ internal static int GetId(RuntimeTypeHandle typeHandle)
             // check whether a corresponding update is required in ClassDataContract.IsNonAttributedTypeValidForSerialization
             private static DataContract CreateDataContract(int id, RuntimeTypeHandle typeHandle, Type type)
             {
-                lock (s_createDataContractLock)
+                DataContract dataContract = s_dataContractCache[id];
```

Comment:
The full framework is using ```Dictionary<...>``` for this fix, which was made by @jcondex. I think this is a targeted fix for the scenario of concurrent deserialization. There was an iteration of this fix which uses locks more properly but that introduces performance regression in common scenarios.